VideoLocker/src/main/java/org/edx/mobile/view/CourseDiscussionResponsesFragment.java

```diff
@@ -148,4 +123,90 @@ public void onClickAddComment(@NonNull DiscussionComment comment) {
+        private boolean isEndorsedFetched = false;
+
+        public CommentsLoader(boolean isQuestionTypeThread) {
+            this.isQuestionTypeThread = isQuestionTypeThread;
+        }
+
+        public void loadComments(@NonNull InfiniteScrollUtils.PageLoadCallback<DiscussionComment> callback) {
+            this.callback = callback;
+            if (isQuestionTypeThread && !isEndorsedFetched) getEndorsedList();
+            getCommentList();
+        }
+
+        /**
+         * Gets the list of endorsed answers for a {@link DiscussionThread.ThreadType#QUESTION}
+         * type discussion thread
+         */
+        protected void getEndorsedList() {
+            if (getEndorsedListTask != null) {
+                getEndorsedListTask.cancel(true);
+            }
+            getEndorsedListTask = new GetCommentListTask(getActivity(),
+                    discussionThread.getIdentifier(), 1, isQuestionTypeThread, true) {
+                @Override
+                public void onSuccess(ThreadComments threadComments) {
+                    isEndorsedFetched = true;
+                    courseDiscussionResponsesAdapter.addToTop(threadComments.getResults());
+                    // If the unendorsed call returned earlier than this one
+                    if (unendorsedComments != null) finalizeCommentLoading();
+                }
+
+                @Override
+                public void onException(Exception ex) {
+                    logger.error(ex);
+                }
+            };
+            getEndorsedListTask.setProgressCallback(null);
+            getEndorsedListTask.execute();
+        }
+
+        protected void getCommentList() {
+            if (getCommentListTask != null) {
+                getCommentListTask.cancel(true);
+            }
+            getCommentListTask = new GetCommentListTask(getActivity(),
+                    discussionThread.getIdentifier(),
+                    nextPage, isQuestionTypeThread, false) {
+                @Override
+                public void onSuccess(final ThreadComments threadComments) {
+                    unendorsedComments = threadComments;
+                    if (isQuestionTypeThread) {
+                        if (isEndorsedFetched) finalizeCommentLoading();
+                    } else {
+                        finalizeCommentLoading();
+                    }
+                }
+
+                @Override
+                public void onException(Exception ex) {
+                    logger.error(ex);
+                }
+            };
+            getCommentListTask.setProgressCallback(null);
+            getCommentListTask.execute();
+        }
+
+        private void finalizeCommentLoading() {
+            ++nextPage;
+            boolean hasMore = unendorsedComments.next != null && unendorsedComments.next.length() > 0;
+            callback.onPageLoaded(unendorsedComments.getResults(), hasMore);
+        }
+
+        public void setNextPage(int page) {
```

Comment:
This can be replaced with a `reset()` method, that resets the whole state instead of just the page:

````java
public void reset() {
    if (getCommentListTask != null) {
        getCommentListTask.cancel(true);
        getCommentListTask = null;
    }
    if (getEndorsedListTask != null) {
        getEndorsedListTask.cancel(true);
        getEndorsedListTask = null;
    }
    unendorsedComments = null;
    isEndorsedFetched = false;
    callback = null;
    nextPage = 1;
}
````